# Домашнее задание 7

### Выполнил Штанько Константин, группа 4

Для начала нам необходимо получить таблицу c данными. Для начала обозначим `расшифровку` каждого столбца для понимания того, что мы анализируем. Давайте выведем ее на экран:

In [29]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import OneHotEncoder



data = pd.read_excel('russian_regions.xlsx', sheet_name='Data')

desc = pd.read_excel('russian_regions.xlsx', sheet_name='Description of data')
print("Таблица 1. Описание данных")
print(desc.to_string(index=False))

Таблица 1. Описание данных
                      Переменная                                                                       Описание переменной
                 Q_OF_LIFE_INDEX                                                                    РЕЙТИНГ КАЧЕСТВА ЖИЗНИ
                  AVERAGE_INCOME                                     СРЕДНЕДУШЕВЫЕ ДЕНЕЖНЫЕ ДОХОДЫ НАСЕЛЕНИЯ В МЕС. (руб.)
                             GRP                                                              ВРП НА ДУШУ НАСЕЛЕНИЯ (руб.)
            POP_PER_HOSPITAL_BED                       ЧИСЛЕННОСТЬ НАСЕЛЕНИЯ НА ОДНУ БОЛЬНИЧНУЮ КОЙКУ НА КОНЕЦ ГОДА (чел.)
                  POP_PER_DOCTOR                                                     ЧИСЛЕННОСТЬ НАСЕЛЕНИЯ НА ОДНОГО ВРАЧА
                   SICKNESS_RATE                                                   ЗАБОЛЕВАЕМОСТЬ (на 1000 чел. населения)
       ATMOSPHERIC_AIR_POLLUTION                             ВЫБРОСЫ ЗАГРЯЗНЯЮЩИХ ВЕЩЕСТВ В АТМОСФЕРНЫЙ ВОЗДУХ (

Теперь посмотрим на саму таблицу с данными, в которой содержатся `различные показатели по регионам России`:

In [2]:
data.head()

REGION  Q_OF_LIFE_INDEX  AVERAGE_INCOME       GRP  \
0  Белгородская область           64.149           30778  506420.9   
1      Брянская область           49.343           26585  253100.4   
2  Владимирская область           50.967           23539  300273.6   
3   Воронежская область           59.327           30289  370610.4   
4    Ивановская область           47.754           24503  182398.1   

   POP_PER_HOSPITAL_BED  POP_PER_DOCTOR  SICKNESS_RATE  \
0                 138.5           247.1          709.9   
1                 132.0           246.9          767.4   
2                 119.9           286.4          919.2   
3                 121.9           191.4          526.6   
4                 122.0           226.0          902.5   

   ATMOSPHERIC_AIR_POLLUTION  DISCHARGE_OF_POLLUTED_WASTEWATER  \
0                        175                          0.299145   
1                         40                          0.600000   
2                         42                          0.811475   
3                        104                          0.319892   
4                         22                          0.487805   

   UNEMPLOYMENT_RATE  EMPLOYMENT_RATE  LABOR_RATE  ENTERPRISES_NUM  \
0                4.9         0.225714    0.240952      33992.24750   
1                4.0         0.236190    0.248571      18703.37125   
2                5.6         0.291905    0.300952      30494.11625   
3                4.3         0.284762    0.289524      51627.62375   
4                5.4         0.274286    0.296190      29030.24625   

   ENTERPRISES_NUM_GROWTH_RATE      FEM_WAGE     MALE_WAGE  WAGE_GENDER_GAP  \
0                     1.137930  31105.842284  39966.872605        77.829063   
1                     0.880532  25282.041814  33424.005888        75.640370   
2                     0.902246  30105.172162  39248.843377        76.703336   
3                     0.924481  31127.016566  41131.373061        75.677067   
4                     1.002322  24944.968012  31452.071954        79.311048   

   FEM_SHARE_IN_EMPLOYEES   FEDERAL_OKRUG  
0                52.25989  Центральный ФО  
1                55.08514  Центральный ФО  
2                57.19788  Центральный ФО  
3                54.69041  Центральный ФО  
4                62.41403  Центральный ФО

In [ ]:
data.describe()

Q_OF_LIFE_INDEX  AVERAGE_INCOME           GRP  POP_PER_HOSPITAL_BED  \
count        81.000000       81.000000  8.100000e+01             81.000000   
mean         48.063407    29103.975309  4.277426e+05            121.255556   
std          11.436130    10907.045156  3.088424e+05             21.110880   
min          19.105000    15603.000000  1.148441e+05             76.200000   
25%          41.395000    23270.000000  2.531004e+05            109.600000   
50%          48.063000    25888.000000  3.671382e+05            121.100000   
75%          54.379000    30778.000000  4.533023e+05            131.100000   
max          81.352000    78812.000000  1.900323e+06            225.500000   

       POP_PER_DOCTOR  SICKNESS_RATE  ATMOSPHERIC_AIR_POLLUTION  \
count       81.000000      81.000000                  81.000000   
mean       218.074074     790.765432                 210.703704   
std         40.129954     159.898472                 403.671817   
min        123.100000     428.300000                   1.000000   
25%        193.200000     685.600000                  31.000000   
50%        218.700000     767.400000                  84.000000   
75%        239.100000     905.700000                 204.000000   
max        341.400000    1278.300000                2319.000000   

       DISCHARGE_OF_POLLUTED_WASTEWATER  UNEMPLOYMENT_RATE  EMPLOYMENT_RATE  \
count                         81.000000          81.000000        81.000000   
mean                           0.399817           7.122222         0.283674   
std                            0.292327           4.121104         0.054079   
min                            0.000000           2.600000         0.101905   
25%                            0.130952           5.000000         0.259524   
50%                            0.423313           5.900000         0.283810   
75%                            0.572204           7.700000         0.315714   
max                            1.168305          29.800000         0.465714   

       LABOR_RATE  ENTERPRISES_NUM  ENTERPRISES_NUM_GROWTH_RATE      FEM_WAGE  \
count   81.000000        81.000000                    81.000000     81.000000   
mean     0.305550     48811.850772                     0.896640  34807.349236   
std      0.058050     94022.928565                     0.219001  13177.414658   
min      0.116667      1098.500000                     0.288545  22114.799304   
25%      0.282857     14619.246250                     0.815531  26653.970458   
50%      0.303333     27372.493750                     0.872834  31127.016566   
75%      0.336190     42723.368750                     0.977018  36825.279251   
max      0.546667    779432.118750                     2.354780  87822.617122   

           MALE_WAGE  WAGE_GENDER_GAP  FEM_SHARE_IN_EMPLOYEES  
count      81.000000        81.000000               81.000000  
mean    46325.949121        76.133470               54.903409  
std     17915.619872         7.271893                4.857662  
min     23006.176368        59.464015               39.475590  
25%     34568.902008        72.104932               52.598810  
50%     41616.546397        75.614298               55.041500  
75%     50519.846072        78.612332               57.197880  
max    112419.904787        97.980853               65.233960

Проверим, нет ли `пропущенных значений`:

In [ ]:
data.isna().sum()

REGION                              0
Q_OF_LIFE_INDEX                     0
AVERAGE_INCOME                      0
GRP                                 0
POP_PER_HOSPITAL_BED                0
POP_PER_DOCTOR                      0
SICKNESS_RATE                       0
ATMOSPHERIC_AIR_POLLUTION           0
DISCHARGE_OF_POLLUTED_WASTEWATER    0
UNEMPLOYMENT_RATE                   0
EMPLOYMENT_RATE                     0
LABOR_RATE                          0
ENTERPRISES_NUM                     0
ENTERPRISES_NUM_GROWTH_RATE         0
FEM_WAGE                            0
MALE_WAGE                           0
WAGE_GENDER_GAP                     0
FEM_SHARE_IN_EMPLOYEES              0
FEDERAL_OKRUG                       0
dtype: int64

Пропущенных значений **нет**, следовательно, можно идти дальше.

### **Задание 1:**

Выберем из общей таблицы те столбцы, с которыми мы будем работать (уберем REGION, т.к. можно использовать FEDERAL_OKRUG):

In [77]:
X = data.drop(columns=['Q_OF_LIFE_INDEX', 'REGION'])
y = data['Q_OF_LIFE_INDEX']
X

AVERAGE_INCOME        GRP  POP_PER_HOSPITAL_BED  POP_PER_DOCTOR  \
0            30778   506420.9                 138.5           247.1   
1            26585   253100.4                 132.0           246.9   
2            23539   300273.6                 119.9           286.4   
3            30289   370610.4                 121.9           191.4   
4            24503   182398.1                 122.0           226.0   
..             ...        ...                   ...             ...   
76           30937   332533.2                 104.5           181.7   
77           59774  1088347.4                  84.3           156.8   
78           53783  1577910.3                  87.9           172.2   
79           24696   322720.4                  82.6           276.2   
80           78812  1386085.3                  76.2           132.4   

    SICKNESS_RATE  ATMOSPHERIC_AIR_POLLUTION  \
0           709.9                        175   
1           767.4                         40   
2           919.2                         42   
3           526.6                        104   
4           902.5                         22   
..            ...                        ...   
76          852.9                        113   
77          657.6                         29   
78          755.0                         85   
79          657.2                         19   
80         1278.3                         23   

    DISCHARGE_OF_POLLUTED_WASTEWATER  UNEMPLOYMENT_RATE  EMPLOYMENT_RATE  \
0                           0.299145                4.9         0.225714   
1                           0.600000                4.0         0.236190   
2                           0.811475                5.6         0.291905   
3                           0.319892                4.3         0.284762   
4                           0.487805                5.4         0.274286   
..                               ...                ...              ...   
76                          0.943662                6.0         0.327619   
77                          0.130952                5.6         0.369048   
78                          0.165563                5.5         0.416190   
79                          0.733333                6.3         0.276667   
80                          0.115385                4.4         0.302381   

    LABOR_RATE  ENTERPRISES_NUM  ENTERPRISES_NUM_GROWTH_RATE      FEM_WAGE  \
0     0.240952      33992.24750                     1.137930  31105.842284   
1     0.248571      18703.37125                     0.880532  25282.041814   
2     0.300952      30494.11625                     0.902246  30105.172162   
3     0.289524      51627.62375                     0.924481  31127.016566   
4     0.296190      29030.24625                     1.002322  24944.968012   
..         ...              ...                          ...           ...   
76    0.358571      15085.49375                     1.092044  42265.904622   
77    0.448571       4287.75000                     0.591495  74167.720029   
78    0.411905      15407.99250                     0.943944  71632.502461   
79    0.321429       2804.37500                     0.892545  32697.059710   
80    0.351905       1098.50000                     0.742230  87822.617122   

        MALE_WAGE  WAGE_GENDER_GAP  FEM_SHARE_IN_EMPLOYEES       FEDERAL_OKRUG  
0    39966.872605        77.829063                52.25989      Центральный ФО  
1    33424.005888        75.640370                55.08514      Центральный ФО  
2    39248.843377        76.703336                57.19788      Центральный ФО  
3    41131.373061        75.677067                54.69041      Центральный ФО  
4    31452.071954        79.311048                62.41403      Центральный ФО  
..            ...              ...                     ...                 ...  
76   66846.650664        63.228156                46.95234  Дальневосточный ФО  
77   89531.921709        82.839415                45.96691  Дальневосточный 

Проведем все первоначальные действия с данными: возьмем Q_OF_LIFE_INDEX за целевую переменную и преобразуем качественную переменную в количественную с помощью OneHotEncoder и разделим данные на тестовую и обучающую выборки:

In [78]:
X = pd.get_dummies(X, columns=['FEDERAL_OKRUG'], drop_first=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

Сначала сделаем бэггинг (построим регрессию на индекс качества жизни):

In [79]:
bagg = RandomForestRegressor(random_state=0,n_estimators=100)
bagg.fit(X_train, y_train)
bagg_pred = bagg.predict(X_test)
bagg_mse = mean_squared_error(y_test, bagg_pred)
bagg_rmse = np.sqrt(bagg_mse)

print(f'Качество бэггинга на тесте:\nMSE: {bagg_mse:.2f}\nRMSE: {bagg_rmse:.2f}')

Качество бэггинга на тесте:
MSE: 44.42
RMSE: 6.66


Теперь сделаем то же самое для бустинга:

In [81]:
grab = GradientBoostingRegressor(random_state=0, n_estimators=100)
grab.fit(X_train, y_train)
grab_pred = grab.predict(X_test)
grab_mse = mean_squared_error(y_test, grab_pred)
grab_rmse = np.sqrt(grab_mse)
print(f'Качество бустинга на тесте:\nMSE: {grab_mse:.2f}\nRMSE: {grab_rmse:.2f}')

Качество бустинга на тесте:
MSE: 38.73
RMSE: 6.22


In [84]:
y_test.describe()

count    17.000000
mean     50.997941
std      12.314794
min      19.105000
25%      47.154000
50%      49.791000
75%      59.037000
max      80.241000
Name: Q_OF_LIFE_INDEX, dtype: float64

Таким образом, можно заметить, что градиентный бустинг справился лучше на тестовой выборке. Однако, в целом, модели показывают не очень хорошие результаты в силу того, что стандартное отклонение индекса счастья от среднего в тестовой выборке составляет 12, а модели в среднем ошибаются на 6 у.е.

### **Задание 2:**

In [97]:
pokazateli = []
for i in np.arange(0.1, 1.1, 0.1):
    maxf = max(1, int(i * X_train.shape[1]))
    randfor = RandomForestRegressor(
        max_features=maxf,
        random_state=0,
        n_estimators=100
    )
    randfor.fit(X_train, y_train)
    predrand = randfor.predict(X_test)
    mserand = mean_squared_error(y_test, predrand)
    rmserand = np.sqrt(mserand)
    pokazateli.append({'maxik': maxf,'MSE': mserand, 'RMSE': rmserand})

In [98]:
print('\n'.join([f"Кол-во признаков: {x['maxik']}, MSE: {x['MSE']:.2f}, RMSE: {x['RMSE']:.2f}" for x in sorted(pokazateli, key=lambda x: x['MSE'])]))

Кол-во признаков: 9, MSE: 38.63, RMSE: 6.22
Кол-во признаков: 18, MSE: 38.78, RMSE: 6.23
Кол-во признаков: 13, MSE: 38.88, RMSE: 6.24
Кол-во признаков: 16, MSE: 41.39, RMSE: 6.43
Кол-во признаков: 11, MSE: 41.59, RMSE: 6.45
Кол-во признаков: 20, MSE: 43.04, RMSE: 6.56
Кол-во признаков: 23, MSE: 44.42, RMSE: 6.66
Кол-во признаков: 6, MSE: 51.42, RMSE: 7.17
Кол-во признаков: 4, MSE: 57.27, RMSE: 7.57
Кол-во признаков: 2, MSE: 75.40, RMSE: 8.68


Получается, что лучше всего справляется модель, которая берет не более 9 признаков для разделения узла. Ее ошибка составляет 6.22 (по RMSE).

In [107]:
best = sorted(pokazateli, key=lambda x: x['MSE'])[0]
print(f'У лучшей модели RandomForest, которую мы подобрали в предыдущем пункте, получились следующие результаты: \nMSE: {best["MSE"]:.2f}\nRMSE: {best["RMSE"]:.2f}')

print(f'\nКачество бэггинга на тесте получилось:\nMSE: {bagg_mse:.2f}\nRMSE: {bagg_rmse:.2f}')
print(f'\nКачество бустинга на тесте:\nMSE: {grab_mse:.2f}\nRMSE: {grab_rmse:.2f}')

У лучшей модели RandomForest, которую мы подобрали в предыдущем пункте, получились следующие результаты: 
MSE: 38.63
RMSE: 6.22

Качество бэггинга на тесте получилось:
MSE: 44.42
RMSE: 6.66

Качество бустинга на тесте:
MSE: 38.73
RMSE: 6.22


Видно, что лучшая модель RandomForest справилась так же хорошо, как и бустинг, только в бустинге я не подбирал оптимальные параметры. Таким образом, вывод заключается в том, что модель здесь работают лучше, чем предсказывающее среднее, однако требуют доработки. К данным больше подошел бустинг, чем бэггинг, так как без настроек он выдает лучшие результаты. При этом, случайный лес смог догнать по качеству бустинг после настройки модели.